src/Npgsql/TypeHandlers/ArrayHandler.cs

```diff
@@ -150,31 +135,159 @@ protected async ValueTask<Array> ReadArray<TAnyElement>(NpgsqlReadBuffer buf, bo
         }
 
         /// <summary>
-        /// Reads an array of element type <typeparamref name="TAnyElement"/> from the given buffer <paramref name="buf"/>.
+        /// Reads a generic list containing elements of type <typeparamref name="TRequestedElement"/> from the given buffer <paramref name="buf"/>.
         /// </summary>
-        protected async ValueTask<List<TAnyElement>> ReadList<TAnyElement>(NpgsqlReadBuffer buf, bool async)
+        protected async ValueTask<List<TRequestedElement>> ReadList<TRequestedElement>(NpgsqlReadBuffer buf, bool async)
         {
             await buf.Ensure(12, async);
             var dimensions = buf.ReadInt32();
 
             if (dimensions == 0)
-                return new List<TAnyElement>();
+                return ElementTypeInfo<TRequestedElement>.EmptyList;
             if (dimensions > 1)
-                throw new NotSupportedException($"Can't read multidimensional array as List<{typeof(TAnyElement).Name}>");
+                throw new NotSupportedException($"Can't read multidimensional array as List<{typeof(TRequestedElement).Name}>");
+
+            var containsNulls = buf.ReadInt32() == 1;
+            if (!ElementTypeInfo<TRequestedElement>.IsReferenceOrNullable && containsNulls)
+                throw new InvalidOperationException(
+                    $"Can't read a non-nullable List<{typeof(TRequestedElement).Name}> if the database array field contains null values. Use List<{typeof(TRequestedElement).Name}?> instead.");
 
-            buf.ReadInt32();  // Has nulls. Not populated by PG?
             buf.ReadUInt32(); // Element OID. Ignored.
 
             await buf.Ensure(8, async);
             var length = buf.ReadInt32();
             buf.ReadInt32(); // We don't care about the lower bounds
 
-            var list = new List<TAnyElement>(length);
+            var list = new List<TRequestedElement>(length);
             for (var i = 0; i < length; i++)
-                list.Add(await _elementHandler.ReadWithLength<TAnyElement>(buf, async));
+                list.Add(await ElementHandler.ReadWithLength<TRequestedElement>(buf, async));
             return list;
         }
 
+        #endregion Read
+
+        #region Static generic caching helpers
+
+        internal static class ElementTypeInfo<TElement>
+        {
+            public static readonly bool IsReferenceOrNullable =
+                !typeof(TElement).IsValueType ||
+                typeof(TElement).IsGenericType && typeof(TElement).GetGenericTypeDefinition() == typeof(Nullable<>);
+
+            public static readonly List<TElement> EmptyList = new List<TElement>(0);
+            // ReSharper disable StaticMemberInGenericType
+            public static readonly Func<ArrayHandler, NpgsqlReadBuffer, bool, ValueTask<Array>> ReadNullableArrayFunc = default!;
+            // ReSharper restore StaticMemberInGenericType
```

Comment:
No, IMO it doesn't.
It warns us, that we'll end up with one instance of `ReadNullableArrayFunc`  per `TElement` which is exactly what we want as we use `ElementTypeInfo<TElement>` as cache for the compiled expression trees. Mind you, that `ElementTypeInfo<TElement>` is nested inside `ArrayHandler` not inside `ArrayHandler<TElement>` so in this case `TElement` might be the element type of any array that the consumer requested.
Perhaps it might be a good idea to rename `TElement` to `TAnyElement` in this case to avoid confusion.
What do you think?

Reply:
Ah, yes, you are right (: